File: 백준/Silver/6603. 로또/로또.py
import sys
input = sys.stdin.readline

def bt(start):
    if len(result)==6:
        print(*result)
        return
    remember = 0
    for i in range(start,num):
        if visited[i] == 0 and remember != lis[i] and lis[i] > remember:
            result.append(lis[i])
            remember = lis[i]
            visited[i] = 1
            bt(i+1)
            result.pop()
            visited[i] = 0
while(True):
    lis = list(map(int,input().split()))
    num = lis[0]
    if num == 0:
        break
    lis = lis[1:]
    result = []
    visited = [0] * (num+1)
    bt(0)
    print()

File: 백준/Silver/6603. 로또/로또.java
import java.io.*;
import java.util.*;

public class Main {
    void bt(int start, ArrayList<Integer> arr, LinkedList<Integer> result, int n) {
        if (result.size() == 6) {
            for (int num : result) {
                System.out.print(num + " ");
            }
            System.out.println();
            return;
        }

        for (int i = start; i < n; i++) {
            result.add(arr.get(i));
            bt(i + 1, arr, result, n);
            result.removeLast();
        }
    }

    public static void main(String[] args) throws IOException {
        Main main = new Main();
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String line;
        while ((line = br.readLine()) != null) {
            StringTokenizer st = new StringTokenizer(line);
            int n = Integer.parseInt(st.nextToken());
            if (n == 0) break;

            ArrayList<Integer> arr = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                arr.add(Integer.valueOf(st.nextToken()));
            }

            LinkedList<Integer> res = new LinkedList<>();
            main.bt(0, arr, res, n);
            System.out.println();
        }
    }
}
